File: src/server/services/ProgressionService.ts
import { Service, OnStart } from "@flamework/core";
import { DataService } from "./DataService/index.js";

@Service({})
export class ProgressionService {
	constructor(private dataService: DataService) {}
	
	getMaxXP(level: number) {
		return level * 2;
	}

	addXP(player: Player, xp: number) {
		const data = this.dataService.get(player);
		if (!data) return;

		const level = data.level;
		const currentXP = data.xp;

		let levelsGained = 0;
		let newXP = currentXP + xp;
		let currentMaxXP = this.getMaxXP(level);

		while (newXP >= currentMaxXP) {
			levelsGained++;
			newXP -= currentMaxXP;
			currentMaxXP = this.getMaxXP(level + levelsGained);
		}

		this.dataService.update(player, (data) => {
			data.level += levelsGained;
			data.xp = newXP;
			data.xpMax = currentMaxXP;
		})

		if (levelsGained > 0) {
			print("Good job!")
		}
	}
}

File: src/server/services/ProgressionService.lua
local ProgressionService = {};

local function getMaxXP(level: number)
	return level * 2;
end

function ProgressionService:init(debugMode)
	if (debugMode) then
		print("<ProgressionService> initialized successfully!")
	end
end

function ProgressionService:ready()
	ProgressionService.cardService = require(script.Parent.CardService);
	ProgressionService.dataService = require(script.Parent.DataService);
end

function ProgressionService:addXP(player, xp)
	local data = ProgressionService.dataService:get(player);
	if (not data) then return end;

	local level = data.level;
	local currentXP = data.xp;
	
	local levelsGained = 0;
	
	local newXP = currentXP + xp;
	local currentMaxXP = getMaxXP(level);
	
	while (newXP >= currentMaxXP) do
		levelsGained += 1;
		newXP -= currentMaxXP;
		currentMaxXP = getMaxXP(level + levelsGained);
	end
	
	ProgressionService.dataService:update(player, function(data)
		data.level = level + levelsGained;
		data.xp = newXP;
		data.xpMax = getMaxXP(data.level);
	end);
	
	if (levelsGained > 0) then
		ProgressionService:TriggerLevelUp(player, levelsGained);
	end
end

function ProgressionService:TriggerLevelUp(player, times)
	for i = 1, times do
		ProgressionService.cardService:showCard(player);
	end
end

return ProgressionService;